### DISCUSSION TITLE
[Messenger] batch handler and delayed flush

### DISCUSSION THREAD

**Discussion by AdamKatzDev** [symfony/symfony#46869] [Q&A] [answered]
We have a need to make requests to external service that has harsh rate limits. So we've decided to not flush batches until some time passes. Lots of issues occurred, some related to Redis Transport, some to Messenger component itself.

But the main question I have is this: is Messenger meant to handle delayed batches? Specifically the following issues made me question this:
-  looks like handler has to flush even when force flag is false because flush with force=true is not called on worker termination. We had to write an event listener for WorkerStoppedEvent to flush our handlers properly.
- looks like flush method is not called when queue is empty (called only after first empty message) which is a problem for this specific case with delayed flushes, a batch might not be flushed for a much longer time than expected.

**Comment by ilya-levin-lokalise** [comment]
We faced a similar situation - we wanted the handler to process a batch of messages OR if there are not enough messages for a batch - process what is there anyways after a short delay. Due to PHP's nature, the only way I can think of implementing it is using `register_tick_function`, which looks hacky and performance-demanding. Another (more straightforward) way is to configure something _outside_ of the application that will be sending signals to the PHP app using some interval - this is cleaner as PHP can handle signals in an async fashion but requires more infra configuration.
@AdamKatzDev let me know if you figured out better solutions :pray:

**Reply by AdamKatzDev** [reply, marked as answer]
We've built an event listener that flushes our handlers with the `false` flag after every `WorkerRunningEvent`. Every batch handler in our case has a timer, when flush is called handlers process current batch if enough time has passed or the batch is full.

The event listener is quite a thing in itself, I'll provide the code just in case someone needs something similar.

src/EventSubscriber/Messenger/FlushBatchesOnWorkerEventsListener.php

```
<?php

declare(strict_types=1);

namespace App\EventSubscriber\Messenger;

use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\Messenger\Event\WorkerMessageReceivedEvent;
use Symfony\Component\Messenger\Event\WorkerRunningEvent;
use Symfony\Component\Messenger\Event\WorkerStoppedEvent;
use Symfony\Component\Messenger\Handler\BatchHandlerInterface;
use Symfony\Component\Messenger\Handler\HandlersLocatorInterface;
use Symfony\Component\Messenger\Worker;

class FlushBatchesOnWorkerEventsListener implements EventSubscriberInterface
{
    /** @var BatchHandlerInterface[][] */
    private array $batchHandlersToFlush = [];

    private \Closure $ackCall;

    public function __construct(private readonly HandlersLocatorInterface $handlersLocator)
    {
    }

    public function onWorkerRunning(WorkerRunningEvent $event): void
    {
        if ($event->isWorkerIdle()) {
            $this->flush(false);
            $this->ack($event->getWorker());
        }
    }

    public function onWorkerStopped(WorkerStoppedEvent $event): void
    {
        $this->flush(true);
        $this->ack($event->getWorker());
    }

    public function onWorkerMessageReceived(WorkerMessageReceivedEvent $event): void
    {
        $messageClass = $event->getEnvelope()->getMessage()::class;

        if (isset($this->batchHandlersToFlush[$messageClass])) {
            return;
        }

        $this->batchHandlersToFlush[$messageClass] = [];

        foreach ($this->handlersLocator->getHandlers($event->getEnvelope()) as $handlerDescriptor) {
            if ($batchHandler = $handlerDescriptor->getBatchHandler()) {
                $this->batchHandlersToFlush[$messageClass][] = $batchHandler;
            }
        }
    }

    private function ack(Worker $worker): void
    {
        if (!isset($this->ackCall)) {
            $this->ackCall = \Closure::bind(static fn() => $worker->ack(), null, $worker);
        }

        ($this->ackCall)();
    }

    private function flush(bool $force): void
    {
        foreach ($this->batchHandlersToFlush as $handlers) {
            foreach ($handlers as $handler) {
                $handler->flush($force);
            }
        }
    }

    public static function getSubscribedEvents(): array
    {
        return [
            WorkerRunningEvent::class         => ['onWorkerRunning'],
            WorkerStoppedEvent::class         => ['onWorkerStopped'],
            WorkerMessageReceivedEvent::class => ['onWorkerMessageReceived']
        ];
    }
}


```

config/services.yaml
```
    App\EventSubscriber\Messenger\FlushBatchesOnWorkerEventsListener:
        arguments:
            $handlersLocator: '@command_bus.messenger.handlers_locator'
```

**Reply by e1sep0** [reply]
Hi @AdamKatzDev, please, could you answer, how to use your listener and configure delay ?
F.E. we want process 30 messages, or if there no any messages during 5 seconds ? Thank you)

**Comment by ilya-levin-lokalise** [comment]
Thanks a lot, @AdamKatzDev we actually found the `WorkerRunningEvent` yesterday as well. Great idea

**Reply by AdamKatzDev** [reply]
Highly advise to use the provided code, it includes a workaround for [this issue](https://github.com/symfony/symfony/issues/49026).
